python/veles/server/conn.py

```diff
@@ -66,6 +77,230 @@ def get_node_norefresh(self, obj_id):
+            raise ObjectGoneError()
+        parent = self.get_node_norefresh(op.parent)
+        if parent.node is None and parent != self.root:
+            raise ObjectGoneError()
+        if parent.id == node.parent.id:
+            return
+        cur = parent
+        while cur.id != NodeID.root_id:
+            if cur.id == node.id:
+                raise ParentCycleError()
+            cur = cur.parent
+        self.db.set_parent(node.id, parent.id, commit=False)
+        node.node.parent = parent.id
+        node.parent = parent
+
+    def _op_set_pos(self, xact, op, node):
+        if node.node is None:
+            raise ObjectGoneError()
+        if op.pos_start == node.pos_start and op.pos_end == node.pos_end:
+            return
+        self.db.set_pos(node.id, op.pos_start, op.pos_end, commit=False)
+        node.node.pos_start = op.pos_start
+        node.node.pos_end = op.pos_end
+
+    def _op_add_tag(self, xact, op, node):
+        if node.node is None:
+            raise ObjectGoneError()
+        if op.tag in node.node.tags:
+            return
+        self.db.add_tag(node.id, op.tag, commit=False)
+        node.node.tags.add(op.tag)
+
+    def _op_del_tag(self, xact, op, node):
+        if node.node is None:
+            raise ObjectGoneError()
+        if op.tag not in node.node.tags:
+            return
+        self.db.del_tag(node.id, op.tag, commit=False)
+        node.node.tags.remove(op.tag)
+
+    def _op_set_attr(self, xact, op, node):
+        if node.node is None:
+            raise ObjectGoneError()
+        if node.node.attr.get(op.key) == op.data:
+            return
+        self.db.set_attr(node.id, op.key, op.data, commit=False)
+        if op.data is None:
+            del node.node.attr[op.key]
+        else:
+            node.node.attr[op.key] = op.data
+
+    def _op_set_data(self, xact, op, node):
+        if node.node is None:
+            raise ObjectGoneError()
+        self.db.set_data(node.id, op.key, op.data, commit=False)
+        xact.set_data(node, op.key, op.data)
+        if op.data is None and op.key in node.node.data:
+            node.node.data.remove(op.key)
+        elif op.data is not None and op.key not in node.node.data:
+            node.node.data.add(op.key)
+
+    def _op_set_bindata(self, xact, op, node):
+        if node.node is None:
+            raise ObjectGoneError()
+        self.db.set_bindata(node.id, op.key, op.start, op.data,
+                            op.truncate, commit=False)
+        old_len = node.node.bindata.get(op.key, 0)
+        if op.truncate:
+            new_len = op.start + len(op.data)
+            change_end = max(old_len, new_len)
+        else:
+            new_len = max(op.start + len(op.data), old_len)
+            change_end = op.start + len(op.data)
+        for sub in node.bindata_subs.get(op.key, ()):
+            if sub.end is not None and sub.end <= op.start:
+                continue
+            if change_end <= sub.start:
```

Comment:
Yes - change_end is computed exactly in this function, and cannot be None (the non-transaction path is sloppier).